cmd/thanos/receive.go

```diff
@@ -131,16 +129,22 @@ func registerReceive(app *extkingpin.App) {
 			AllowOverlappingBlocks: *tsdbAllowOverlappingBlocks,
 		}
 
-		// Local is empty, so try to generate a local endpoint
-		// based on the hostname and the listening port.
+		// Make sure the both localEndpoint and (hashringsFilePath/hashringsFileContent) are not empty
+		if *localEndpoint == "" && *hashringsFilePath == "" && *hashringsFileContent == "" {
+			return errors.New("both the receive.local-endpoint and (receive.hashrings-file/receive.hashrings) are empty; see https://thanos.io/tip/components/receive.md for details.")
+		}
+
+		// Default behavior for receiver is both ingestion and distribution
+		receiverMode := receive.DUAL_MODE
```

Comment:
I find this abstraction a bit more confusing than it is helpful as it forces us to have lots of checks in the code like "if this but not this" or "if not that but yes this". I think it would be more helpful to have one single flag that says "i run a TSDB". That is really the only thing we care about. Even if there is no hashring file configured, the ingestor mode MUST run a gRPC server because this mode might be at the bottom of a routing tree and know nothing about the distributor nodes that are forwarding traffic to it. This means that there is no difference at all between "dual mode" and "ingestor"

Reply:
Okay, so I think, we can introduce a flag, `disableTSDB`, which would instruct the receiver to **only route data**, and not ingest data. 
So this would be the following behavior of the receiver - 

* **Both Ingestion and Distribution** - Default behavior of receiver, works as it used to, no extra flags.
* **Only Ingestion** - Just don't pass hash ring file path/content, the receiver would just ingest data.
* **Only Distribution** - We can create a flag, `disableTSDB`, which would stop creating a local tsdb for ingesting data.

Thoughts?

Reply:
I don't see a need for this flag. We can unambiguously infer everything we need to know with the presence/absence of the current flags:

1. Hashring info != "" && local endpoint == "": means only in distribution mode, i.e. don't start a TSDB
2. Everything else means start a regular receiver with TSDB, http server, and a gRPC server

This covers all of the possible cases we could want, does not require any new flags, and keeps things simpler IMO

Reply:
Sure, we can configure this decision internally also, Thanks!

Reply:
In the distribution mode, do we need to start the gRPC server? 

Also, can you list out which parts are enabled in the ingestion / distribution mode
cc @squat

Reply:
Sounds great!